File: Design Patterns/Iterator Pattern/iterator.ts
// The below code demonstrates Iterator pattern.
// It defines a NumberCollection class that holds an array of numbers,
// and a NumberIterator class that provides methods to iterate over the collection.
// The iterator allows sequential access to the elements without exposing the underlying structure.

class NumberCollection {
  private numbers: number[] = [1, 2, 3, 4, 5];

  public getIterator(): NumberIterator {
    return new NumberIterator(this.numbers);
  }
}

class NumberIterator {
  private index: number = 0;

  constructor(private numbers: number[]) {}

  public hasNext(): boolean {
    return this.index < this.numbers.length;
  }

  public next(): number {
    return this.numbers[this.index++];
  }
}

// Usage
const collection = new NumberCollection();
const iterator = collection.getIterator();

while (iterator.hasNext()) {
  console.log(iterator.next());
}

File: Design Patterns/Iterator Pattern/iterator.py
# This code shows the Iterator pattern.
# It defines a NumberCollection class that holds a list of numbers,
# and a NumberIterator class that provides methods to iterate over the collection.
# The iterator allows sequential access to the elements without exposing the underlying structure.

class NumberCollection:
    def __init__(self):
        self.numbers = [1, 2, 3, 4, 5]

    def get_iterator(self):
        return NumberIterator(self.numbers)

class NumberIterator:
    def __init__(self, numbers):
        self.numbers = numbers
        self.index = 0

    def has_next(self):
        return self.index < len(self.numbers)

    def next(self):
        if self.has_next():
            number = self.numbers[self.index]
            self.index += 1
            return number
        else:
            raise StopIteration()

# Usage
collection = NumberCollection()
iterator = collection.get_iterator()

while iterator.has_next():
    print(iterator.next())